API.md

```diff
@@ -0,0 +1,319 @@
+
+
+```typescript=
+export default class Plugin extends LitElement {
+  docs: Record<string, XMLDocument> = {};
+  doc?: XMLDocument;
+  docName?: string;
+  editable: string[] = [
+    'cid',
+    'icd',
+    'iid',
+    'scd',
+    'sed',
+    'ssd',
+  ];
+  history: LogEntry[];
+  editCount: number = 0;
+  plugins: { menu: Plugin[], editor: Plugin[] }[];
+  locale: string = 'en';
+}
+
+/** Helper types exported by OpenSCD core **/
+
+type LogEntry = {
+  redo: Edit; // `Edit` defined below
+  undo: Edit;
+  title?: string
+}
+
+type Plugin = {
+  name: string;
+  translations?: Record<string, string>;
+  src: string;
+  icon: string;
+  requireDoc?: boolean; // disable plugin if no doc is opened
+}
+```
+
+### `docs`
+
+`docs` is set to an object mapping `string` keys (document names) to `XMLDocument` values.
+
+### `docName`
+The name of the `XMLDocument` currently being edited.
+
+### `doc`
+The `XMLDocument` currently being edited.
+
+### `editable`
+Filename extensions of user-editable documents.
+
+### `history`
+History of edits done to `doc`.
+
+### `editCount`
+
+Index of the current log entry in the `history`. Is incremented with every new edit and updated on undo/redo. 
+
+### `plugins`
+
+Arrays of `Plugin` objects describing the currently loaded `menu` and `editor` plugins, respectively.
+
+### `locale`
+
+Selected language (IETF language tag).
+
+## Communicating user intent to OpenSCD core
+
+Plugins communicate user intent to OpenSCD core by dispatching the following [custom events](https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent):
+
+### `EditEventV2`
+
+The **edit event** allows a plugin to describe the changes it wants to make to the current `doc`.
+
+```typescript=
+export type EditDetailV2<E extends Edit = Edit> = {
+  edit: E;
+  title?: string;
+  squash?: boolean;
+}
+
+export type EditEventV2<E extends Edit = Edit> = CustomEvent<EditDetailV2<E>>;
+
+export type EditEventOptions = {
+  title?: string;
+  squash?: boolean;
+}
+
+export function newEditEventV2<E extends Edit>(edit: E, options: EditEventOptions): EditEventV2<E> {
+  return new CustomEvent<E>('oscd-edit-v2', {
+    composed: true,
+    bubbles: true,
+    detail: {...options, edit},
+  });
+}
+
+declare global {
+  interface ElementEventMap {
+    ['oscd-edit-v2']: EditEventV2;
+  }
+}
+```
+
+It's `title` property is a human-readable description of the edit, and `squash` is a boolean indicating whether the edit should be merged with the previous edit in the history.
+
+#### `Edit` type
+
+The `EditDetailV2` defined above contains an `edit` of this type:
+
+```typescript=
+export type Edit = Insert | SetAttributes | Remove | Edit[];
+```
+
+This means that a single edit can either consist in a sequence of other edits or in one of the following atomic edit types:
```

Comment:
It's a bit of an edge case, but changing `textContent` is quite awkward at the moment.
Is it worth having an API for this? I am happy if you decide not to.

Reply:
Would you then expect a new edit type `SetTextContent`?

Reply:
Yes please.

Reply:
Great point, I'll add this to the proposal.